### PR TITLE
🚀 Update variables reconciliation logic to reduce number of API calls

### PR DESCRIPTION
**Pull request by arybolovlev**
### Description

This PR introduces changes to variables reconciliation in the workspace controller

- Extend `status` with variables status:

```yaml
status:
  variables:
    - category: terraform
      id: var-a1hnPzjknedvkoky
      name: username_length
      valueID: 7269bfbb3a64c5a26effe18028e5362f989cf71905038f90da60eefa619b8946
      versionID: 951e711fc7b19146cafedeefb9b092d75c0497420876bef00fa17d57f7b27e20
```

`status.variables.[*].versionID` and `status.variables.[*].valueID` are hashes of a variable on the HCP Terraform and HCP Terraform Operator sides, respectively. Tracking these values allows variables to be updated when they have changed. Previously, sensitive variables were constantly updated during each reconciliation due to their nature, resulting in unnecessary API calls.

#### Tests

- https://github.com/hashicorp/terraform-cloud-operator/actions/runs/8737319244
- https://github.com/hashicorp/terraform-cloud-operator/actions/runs/8737319976

### Usage Example

N/A.

### References

N/a.

### Community Note

* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### CODE REVIEW

File: controllers/workspace_controller_variables.go [hunk 1]
```diff
@@ -5,185 +5,162 @@ package controllers
...
+func variableValueID(v tfc.Variable) string {
+	hash := sha256.New()
+	hv := appv1alpha2.Variable{
+		Name:        v.Key,
+		Description: v.Description,
+		HCL:         v.HCL,
+		Sensitive:   v.Sensitive,
+		Value:       v.Value,
 	}
+	hash.Write([]byte(fmt.Sprintf("%+v", hv)))
```
Comment (alexsomesan) [inline controllers/workspace_controller_variables.go:29]: Nothing wrong with this approach, but I think `fmt.Sprintf` doesn't offer any output stability guarantees.
One simple way serialise the struct to a guaranteed stable output is https://pkg.go.dev/encoding/gob

To use it, you would do this:
```suggestion
	gob.NewEncoder(hash).Encode(hv)
```

No big deal if we keep Sprintf, but just wanted to mention this option.

File: controllers/workspace_controller_variables.go [hunk 2]
```diff
@@ -224,64 +201,50 @@ func getWorkspaceVariablesByCategory(workspaceVariables []*tfc.Variable, categor
...
-	updateVariables := getVariablesToUpdate(instanceVariables, workspaceVariables)
-	if len(updateVariables) > 0 {
-		w.log.Info("Reconcile Variables", "msg", fmt.Sprintf("updating %d %s variables in the workspace ID %s", len(updateVariables), category, workspaceID))
-		err := r.updateWorkspaceVariables(ctx, w, updateVariables)
-		if err != nil {
-			return err
+	w.log.Info("Reconcile Variables", "msg", fmt.Sprintf("there are %d %s variables in spec", len(specVariables), category))
+	w.log.Info("Reconcile Variables", "msg", fmt.Sprintf("there are %d %s variables in workspace", len(workspaceVariables), category))
+
+	// Let's consider spec and workspace variables as two sets.
```
Comment (alexsomesan) [inline controllers/workspace_controller_variables.go:226]: 🆒

**Review by alexsomesan — APPROVED** [review]
Awesome work, love it!

File: controllers/workspace_controller_variables.go (hunk 1 above)
Comment (arybolovlev) [reply controllers/workspace_controller_variables.go:29]: Thanks, I didn't know about `gob`. Switched to it.

**Pull request merged**